polaris/utils/v2_manifest.py

```diff
@@ -0,0 +1,71 @@
+from hashlib import md5
+import os
+import pyarrow as pa
+import pyarrow.parquet as pq
+
+# PyArrow table schema for the V2 Zarr manifest file
+ZARR_MANIFEST_SCHEMA = pa.schema([("path", pa.string()), ("checksum", pa.string())])
+
+
+def generate_zarr_manifest(zarr_root_path: str, output_dir: str):
+    """
+    Entry point function which triggers the creation of a Zarr manifest for a V2 dataset.
+
+    Parameters:
+        zarr_root_path: The path to the root of a Zarr archive
+        output_dir: The path to the directory which will hold the generated manifest
+    """
+
+    zarr_manifest_path = f"{output_dir}/zarr_manifest.parquet"
+
+    with pq.ParquetWriter(zarr_manifest_path, ZARR_MANIFEST_SCHEMA) as writer:
+        recursively_build_manifest(zarr_root_path, writer, zarr_root_path)
+
+    return zarr_manifest_path
+
+
+def recursively_build_manifest(dir_path: str, writer: pq.ParquetWriter, zarr_root_path: str) -> str:
+    """
+    Recursive function that traverses a Zarr archive to build a V2 manifest file.
+
+    Parameters:
+        dir_path: The path to the current directory being processed in the archive
+        writer: Writer object for incrementally adding rows to the manifest Parquet file
+        zarr_root_path: The root path which triggered the first recursive call
+    """
+
+    # Get iterator of items located in the directory at `dir_path`
+    with os.scandir(dir_path) as it:
```

Comment:
We used `os.walk` in the original implementation and it does return a generator that yields a tuple for each directory level, but the `files` item in the tuple is an in-memory array. If there is a directory with millions of files/chunks, it could lead to issues with memory (though probably unlikely). 

We switched to `os.scandir` because each item in the directory is yielded individually.

Reply:
That's fair. I guess in the end that will depend on whether the Zarr structure is broader or deeper. I'd be curious to benchmark each approach once we get an actual XL dataset.